File: yggdrasil/examples/maxMsg/src/maxMsgSrv.c
#include "YggInterface.h"
#include <stdio.h>


int main(int argc, char *argv[]) {  

    printf("maxMsgSrv(C): Hello!\n");
    yggRpc_t rpc = yggRpcServer("maxMsgSrv", "%s", "%s");
    size_t input_size = YGG_MSG_BUF;
    char *input = (char*)malloc(input_size);
    //char input[YGG_MSG_BUF];

    while (1) {
      // Reset to size of buffer if not all utilized
      if (input_size < YGG_MSG_BUF)
	input_size = YGG_MSG_BUF;
      
      int ret = rpcRecvRealloc(rpc, &input, &input_size);
      if (ret < 0)
        break;
      printf("maxMsgSrv(C): rpcRecv returned %d, input (size=%lu) %.10s...\n",
	     ret, input_size, input);
      ret = rpcSend(rpc, input, input_size);
      if (ret < 0) {
        printf("maxMsgSrv(C): SEND ERROR\n");
        break;
      }
    }

    printf("maxMsgSrv(C): Goodbye!\n");
    free(input);
    return 0;
}

File: yggdrasil/examples/maxMsg/src/maxMsgSrv.f90
program main
  use fygg

  type(yggcomm) :: rpc
  integer(kind=c_size_t) :: input_size
  type(yggchar_r) :: input
  logical :: ret

  print *, "maxMsgSrv(F): Hello!"
  rpc = ygg_rpc_server("maxMsgSrv", "%s", "%s")

  do while (.true.)
     ! Reset to size of buffer if not all utilized
     input_size = size(input%x)

     ret = ygg_recv_var_realloc(rpc, [yggarg(input), yggarg(input_size)])
     if (.not.ret) exit
     print *, "maxMsgSrv(F): rpcRecv returned ", ret, ", input (size=", &
          input_size, ") ", input%x(1:10)
     ret = ygg_send_var(rpc, [yggarg(input), yggarg(input_size)])
     if (.not.ret) then
        print *, "maxMsgSrv(F): SEND ERROR"
        exit
     end if

  end do

  print *, "maxMsgSrv(F): Goodbye!"

end program main
